server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -1297,16 +1299,16 @@ public void onSuccess(String source) {
 
                                     if (mode == Mode.LEADER) {
                                         final ClusterState state = getLastAcceptedState(); // committed state
-                                        if (electionQuorumContainsLocalNode(state) == false) {
+                                        if (localNodeMayWinElection(state)) {
+                                            scheduleReconfigurationIfNeeded();
+                                        } else {
                                             final List<DiscoveryNode> masterCandidates = completedNodes().stream()
                                                 .filter(DiscoveryNode::isMasterNode)
-                                                .filter(node -> electionQuorumContains(state, node))
+                                                .filter(node -> nodeMayWinElection(state, node))
                                                 .collect(Collectors.toList());
                                             if (masterCandidates.isEmpty() == false) {
                                                 abdicateTo(masterCandidates.get(random.nextInt(masterCandidates.size())));
                                             }
```

Comment:
I would feel better about calling `scheduleReconfigurationIfNeeded()` in the else branch here, even if it turns out to be a NOOP. Otherwise the interaction between what the reconfigurator does and the logic here is too subtle.